go-controller/pkg/ovn/address_set.go

```diff
@@ -0,0 +1,234 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	kapi "k8s.io/api/core/v1"
+	"k8s.io/klog"
+)
+
+type addressSet struct {
+	sync.Mutex
+
+	name      string
+	hashName  string
+	ports     map[string]net.IP // logical port :: pod IP
+	destroyed bool
```

Comment:
why do we need this?

Reply:
This is so that someone who holds on to a copy of an `addressSet` rather than re-looking-it-up every time they need it can tell (after locking it) whether the `addressSet` has been deleted from the namespace.

It serves the same function as the weird locking/unlocking in `getNamespaceLocked`, except with the locking and the checking being done in a different place. And it's bad that we do the same thing in to two different ways in two very similar pieces of code. We should settle on one or the other system...

Reply:
iiuc this is the result of the double mutex, and releasing the plural lock while waiting for the singular lock. If this is the same concept for a map of address sets with it's own lock trying to lock a single address set, then we should use the same method we are using in the namespace code and not use a destroyed attribute on the address set itself.

Note I also think the namespace double locking is not great, but I don't have a better solution yet. There are concurrent sync maps in golang, but I saw those also have some pitfalls, so not sure those would help. I did want to to test it out though.

Reply:
@trozet I reworked this to remove destroyed, and make sure that callers always hold the lock. It now more-or-less follows what namespace code does.